### ISSUE TITLE
Can not auto discover keeper hostname

### ISSUE THREAD

**Issue by RekGRpth** [citusdata/pg_auto_failover#448]
Can not auto discover keeper hostname
because of, function `inet_addr` in file `ipaddr.c` requires

> Internet host address IPv4 in numbers-and-dots notation

as argument, but not hostname! (as it passed in function `discover_hostname` in file `cli_create_drop_node.c`)

**Comment by RekGRpth** [comment]
how about such solution?
```diff
diff --git a/src/bin/pg_autoctl/ipaddr.c b/src/bin/pg_autoctl/ipaddr.c
index d1c34d1..d9b8a3d 100644
--- a/src/bin/pg_autoctl/ipaddr.c
+++ b/src/bin/pg_autoctl/ipaddr.c
@@ -49,6 +49,7 @@ fetchLocalIPAddress(char *localIpAddress, int size,
        struct sockaddr_in name;
        socklen_t namelen = sizeof(name);
        int err = -1;
+       struct hostent *he;
 
        struct sockaddr_in serv;
 
@@ -63,7 +64,14 @@ fetchLocalIPAddress(char *localIpAddress, int size,
 
        memset(&serv, 0, sizeof(serv));
        serv.sin_family = AF_INET;
-       serv.sin_addr.s_addr = inet_addr(serviceName);
+       if((he = gethostbyname2(serviceName, AF_INET)))
+       {
+               serv.sin_addr.s_addr = ((struct in_addr *)he->h_addr)->s_addr;
+       }
+       else
+       {
+               serv.sin_addr.s_addr = inet_addr(serviceName);
+       }
        serv.sin_port = htons(servicePort);
 
        err = connect(sock, (const struct sockaddr *) &serv, sizeof(serv));
```

**Comment by DimCitus** [comment]
Hi @RekGRpth ; it is very hard for me to understand what you're trying to solve here. Can you please include enough information in your ticket? The basics for bug reports are always the same:

  1. tell us what you did (e.g. copy/paste a `pg_autoctl` command line)
  2. tell us what you obtained (e.g. copy/paste the output of the command line, and then point to the exact part that is a problem for you)
  3. tell us how what you obtained is different from what you expected

Without that I can't help you. And this issue is missing all the 3 points, jumping to a reaction to something only you saw and know about. Help me help you?

**Comment by RekGRpth** [comment]
1) I just try register new keeper without specifying --hostname option

**Comment by DimCitus** [comment]
And how I am supposed to relate that with the rest of the ticket. What is the ERROR MESSAGE that you're seeing, or the hostname that has been picked, what is the hostname that you expected to be discovered, etc?

**Comment by RekGRpth** [comment]
2. I obtaned following
```sh
pg_autoctl -vvv create postgres --server-cert /etc/certs/cert.pem --server-key /etc/certs/key.pem --ssl-ca-file /etc/certs/ca.pem --formation default --monitor 'postgres://autoctl_node@pg_auto_failover0:5432/pg_auto_failover?sslmode=prefer' --name pg
_auto_failover1 --replication-quorum false --skip-pg-hba --ssl-mode prefer
10:14:20 3457 TRACE cli_common.c:669: --server-cert /etc/certs/cert.pem
10:14:20 3457 TRACE cli_common.c:676: --server-key /etc/certs/key.pem
10:14:20 3457 TRACE cli_common.c:655: --ssl-ca-file /etc/certs/ca.pem
10:14:20 3457 TRACE cli_common.c:245: --formation default
10:14:20 3457 TRACE cli_common.c:272: --monitor postgres://autoctl_node@pg_auto_failover0:5432/pg_auto_failover?sslmode=prefer
10:14:20 3457 TRACE cli_common.c:229: --name pg_auto_failover1
10:14:20 3457 TRACE cli_common.c:313: --replication-quorum false
10:14:20 3457 TRACE cli_common.c:213: --skip-pg-hba
10:14:20 3457 TRACE cli_common.c:685: --ssl-mode prefer
10:14:20 3457 TRACE config.c:166: SetConfigFilePath: "/var/lib/postgresql/.config/pg_autoctl/var/lib/postgresql/pg_data/pg_autoctl.cfg"
10:14:20 3457 TRACE config.c:194: SetStateFilePath: "/var/lib/postgresql/.local/share/pg_autoctl/var/lib/postgresql/pg_data/pg_autoctl.state"
10:14:20 3457 TRACE config.c:209: SetKeeperStateFilePath: "/var/lib/postgresql/.local/share/pg_autoctl/var/lib/postgresql/pg_data/pg_autoctl.init"
10:14:20 3457 TRACE config.c:235: SetPidFilePath: "/tmp/pg_autoctl/var/lib/postgresql/pg_data/pg_autoctl.pid"
10:14:20 3457 TRACE pgsetup.c:1602: pgsetup_validate_ssl_settings
10:14:20 3457 WARN  ipaddr.c:91: Failed to connect: Invalid argument
10:14:20 3457 FATAL cli_create_drop_node.c:1277: Failed to find a local IP address, please provide --hostname.
10:14:20 3457 FATAL cli_create_drop_node.c:1206: Failed to auto-detect the hostname of this machine, please provide one via --hostname
10:14:20 3457 TRACE lock_utils.c:156: ipcrm -s 65551
```

**Comment by RekGRpth** [comment]
I mentioned above the reason and solution of this error

**Comment by RekGRpth** [comment]
I updated description of issue

**Comment by DimCitus** [comment]
Thanks for providing necessary information. I now see the same problem as you're seeing, I believe. Your monitor URI contains a hostname rather than an IP address, and so when we use `hostname_from_uri`  in `check_or_discover_hostname` and then give the parsed monitor hostname to `discover_hostname` then to `fetchLocalIPAddress` ; then we have a problem with using `inet_addr` which expects an ipv4 dotted address.

You can use the following command to do the hostname lookup dance already, and see how to play with the other functions and code in `ipaddr.c`:

```
PG_AUTOCTL_DEBUG=1 ./src/bin/pg_autoctl/pg_autoctl do show lookup pg_auto_failover0
```

I believe we should use `getaddrinfo` instead of `inet_addr` in order to be compatible with ipv4 dotted notation, ipv6 notation, and hostnames. Also see the [man page for gethostbyname2](https://man7.org/linux/man-pages/man3/gethostbyname2.3.html):

>        The gethostbyname*(), gethostbyaddr*(), herror(), and hstrerror() functions are obsolete.  Applications should use getaddrinfo(3), getnameinfo(3), and gai_strerror(3) instead.

Do you want to work on a proper PR to solve the problem using `getaddrinfo` here?

**Comment by RekGRpth** [comment]
> Do you want to work on a proper PR to solve the problem using getaddrinfo here?

I use my solution already

**Comment by RekGRpth** [comment]
> believe we should use getaddrinfo

Yes, I solved problem by simplest way, not most right

**Comment by DimCitus** [comment]
You solved the problem in a way that only works in your own setup. To be fair, the problem that we have now is because the code I wrote only works in a limited setup too. So I don't want to change which specific setup is supported by default, but rather support both setups, and also add ipv6 at the same time.

**Comment by handle2001** [comment]
If I understand what fetchLocalIPAddress is being used for, it's doing the second half of the pg_hba-style host-based authentication process. To wit:

> If a host name is specified (anything that is not an IP address range or a special key word is treated as a host name), that name is compared with the result of a reverse name resolution of the client's IP address (e.g., reverse DNS lookup, if DNS is used). Host name comparisons are case insensitive. If there is a match, then a forward name resolution (e.g., forward DNS lookup) is performed on the host name to check whether any of the addresses it resolves to are equal to the client's IP address. If both directions match, then the entry is considered to match.
_(from https://www.postgresql.org/docs/9.1/auth-pg-hba-conf.html)_

Emulation of this entire authentication process appears to happen within the discover_hostname() function in cli_create_drop_node.c, though I see that fetchLocalIPAddress() is used in a few other cases as well to get the local IP in string format. If these observations are correct, I do believe that getaddrinfo() and getnameinfo() could be used in both cases and remove some of the complexity in these functions while natively handling IPv6 also if this is desired, though maybe wrapping the IPv6 stuff in an #ifdef is an exercise for later. Additionally, getaddrinfo() and getnameinfo() will use the host's configured DNS resolver without having to rely on hardcoding Google's IP, and also permits to use private DNS resolvers in the case that speaking with the Internet is undesirable. @RekGRpth is correct that gethostbyname() is simple but it's an obsolete function and creates problems in threaded applications.

I've forked the repo and I'm going to noodle on this a bit.

**Comment by handle2001** [comment]
I think something like this would work https://pastebin.com/ZyduUE5e.

**Comment by DimCitus** [comment]
Hi @handle2001 ; thanks for your contribution! Please see about preparing a PR yes, if you're up for it, that would be awesome. Here's a brain dump of what I think is a good way to approach our situation now, after reading your reaction here and your pseudo-code:

  1. In a Postgres node, still connect to the monitor to see which interface/IP is used, using `getaddrinfo` instead of the current call to `inet_addr`
  2. In a monitor node, first call `gethostname` and then when the result address type from `ip_address_type()` is `IPTYPE_NONE` then call `findHostnameLocalAddress` on the hostname we got, and if we can trace it to a local IP address from one of the interfaces then use the hostname,
  3. In a monitor node, if we fail to get an hostname that resolves to an IP address that is found on the host itself, then attempt to contact an external service to determine an IP address to use, as before.

What do you think?

**Comment by DimCitus** [comment]
Hi @RekGRpth ; I believe your issue is now fixed with the merge of #479. You may have a look at the changes if you want to, to have an idea of the amount of work this required so that we fix the issue for every one. I'm closing this issue how, if you still have problem, please consider re-opening with more information.
